### ISSUE TITLE
Combining singletons with :kind! has worse UX under singletons-2.7

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/singletons#445]
I was surprised to see this in a recent GHCi session with `singletons-2.7`:

```
λ> :kind! Map IdSym0 '[1,2,3]
Map IdSym0 '[1,2,3] :: [Nat]
= 1 :@#@$$$ '[2, 3]
```

Compare this to a session with `singletons-2.6`:

```
λ> :kind! Map IdSym0 '[1,2,3]
Map IdSym0 '[1,2,3] :: [Nat]
= '[1, 2, 3]
```

Why is `singletons-2.7`'s `:kind!` output so much uglier? Ultimately, it's due to #424. Among other things, #424 slightly changed the way fully saturated defunctionalization symbols are used. Before, `singletons` would have generated the following defunctionalization-related code for `(:)`:

```hs
data (:@#@$$) :: a -> [a] ~> [a]
type x :@#@$$$ xs = x ': xs

type instance Apply ((:@#@$$) x) xs = x ': xs
```

But after #424, `singletons` now generates this:

```hs
data (:@#@$$) :: a -> [a] ~> [a]
type x :@#@$$$ xs = x ': xs

type instance Apply ((:@#@$$) x) xs = x :@#@$$$ xs
```

Notice how the right-hand side of the `type instance` is now `x :@#@$$$ xs`, not `x ': xs`. Since `:kind!` only expands type family applications, not type synonym applications, this means that you end up with `1 :@#@$$$ '[2, 3]` as the result of running `:kind! Map IdSym0 '[1,2,3]`. (For reasons not entirely clear to me, you only wind up with one occurrence of the `(:@#@$$$)` type synonym rather than, say, `1 :@#@$$$ 2 :@#@$$$ 3 :@#@$$$ '[]`.)

When writing #424, I made this change since it made the code for generating defunctionalization symbols for uniform. In retrospect, however, it appears that there is a good reason to avoid doing so. I think we should implement a fix for this issue by doing one of the following:

1. Make fully saturated defunctionalization symbols like `(:@#@$$$)` type families, not type synonyms, so that `:kind!` expands them, or
2. Go back to generating `type instance Apply ((:@#@$$) x) xs = x ': xs` instead.

(1) is arguably the cleaner fix, but it would be a breaking change to implement it—see [this part](https://github.com/goldfirere/singletons/blob/f6c4223d3674ea323346ecb03b19c389bac2cdc8/src/Data/Singletons/Promote/Defun.hs#L788-L869) of `Note [No SAKs for fully saturated defunctionalization symbols]`. (2), on the other hand, could be implemented without any user-facing changes.

**Comment by goldfirere** [comment]
I vote (1). It's more uniform. It's helpful to link to that Note, which shows why (1) is potentially painful. But the example there uses an overlapping pattern, which is officially not supported by `singletons`. One could argue that the fact that the overlapping pattern is accepted is a bug. Choosing (1) fixes that bug. :)

I don't feel strongly, though.

It's also worth noting that the behavior of `:kind!` is quite arbitrary and was chosen as a compromise. It is conceivable that the real fix is improving `kind!` to treat type synonyms and type families uniformly, by expanding them all. It's not clear to me whether that would be an improvement or not within GHC.

**Comment by RyanGlScott** [comment]
Hm... it's worth noting that (2) won't completely fix this issue. Observe that the following GHCi session occurs on both `singletons-2.6` and `-2.7`:

```
λ> ; $(promote [d| unit = () |])
λ> :kind! Unit
Unit :: ()
= Tuple0Sym0
```

What's more, (2) won't change this, since `Tuple0Sym0` is a fully saturated defunctionalization symbol and therefore has no `Apply` instance. I can't see a way to fix this without (1).

> It's also worth noting that the behavior of :kind! is quite arbitrary and was chosen as a compromise. It is conceivable that the real fix is improving kind! to treat type synonyms and type families uniformly, by expanding them all. It's not clear to me whether that would be an improvement or not within GHC.

Indeed, there is a hypothetical option (3) in which a proposal like [`:kind!!`](https://github.com/ghc-proposals/ghc-proposals/pull/79) is implemented, and we simply tell people to use `:kind!!` instead of `:kind!`. But that proposal has been abandoned with no signs of being revived.

-----

In short, I think I'm just going to be bold and implement (1). It's more uniform, will fix both bugs documented in this issue, and will let us rip off the bandage that is `Note [No SAKs for fully saturated defunctionalization symbols]`.

**Comment by RyanGlScott** [comment]
Eek. There's another downside to (1): it makes it far easier to trigger a reduction stack overflow. Here is a fairly innocent program that compiles without issue on `singletons-2.7`:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE EmptyCase #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE StandaloneKindSignatures #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
module Bug where

import Data.Singletons.Prelude
import Data.Singletons.TH
import qualified GHC.TypeLits as Lit

$(promote [d| data Nat = Z | S Nat deriving (Eq, Ord) |])

type Lit :: Lit.Nat -> Nat
type family Lit n where
  Lit 0 = Z
  Lit n = S (Lit (n Lit.- 1))
$(genDefunSymbols [''Lit])

$(promoteOnly [d|
  evenb :: Nat -> Bool
  evenb Z = True
  evenb (S Z) = False
  evenb (S (S n)) = evenb n

  filterEvenGt7 :: [Nat] -> [Nat]
  filterEvenGt7 = filter (\x -> evenb x && x > lit 7)
  |])

testFilterEvenGt7 :: FilterEvenGt7 (Map LitSym0 '[5,2,6,19,129]) :~: '[]
testFilterEvenGt7 = Refl
```

With #446, however, this fails to compile with GHC's default reduction depth:

```
[1 of 1] Compiling Bug              ( Bug.hs, interpreted )

Bug.hs:38:21: error:
    • Reduction stack overflow; size = 201
      When simplifying the following type:
        Apply
          EvenbSym0
          ('S ('S ('S ('S ('S ('S ('S ('S ('S ('S ('S ('S ('S 'Z)))))))))))))
      Use -freduction-depth=0 to disable this check
      (any upper bound you could choose might fail unpredictably with
       minor updates to GHC, so disabling the check is recommended if
       you're sure that type checking should terminate)
    • In the expression: Refl
      In an equation for ‘testFilterEvenGt7’: testFilterEvenGt7 = Refl
   |
38 | testFilterEvenGt7 = Refl
   |                     ^^^^
```

I have to set `-freduction-depth` to at least `222` in order to make this compile now. This leads me to believe that we should implement both (1) _and_ (2). Although (2) isn't strictly necessary to fix the `:kind!` issue if (1) is implemented, (2) makes evaluating any unsaturated defunctionalization symbol take one less reduction step since it skips over the fully saturated symbol. As a consequence, the program above continues to compile without having to change `-freduction-depth`.
